clang/test/CodeGen/alignment.c

```diff
@@ -45,7 +45,7 @@ void test3(packedfloat3 *p) {
   *p = (packedfloat3) { 3.2f, 2.3f, 0.1f };
 }
 // CHECK: @test3(
-// CHECK: store <4 x float> {{.*}}, align 4
+// CHECK: store <3 x float> {{.*}}, align 4
```

Comment:
Is the backend able to lower this a a single store? Or will it get split into 2 or more stores? As far as I know the X86 backend will split it into 2 stores, a 64 bit and a 32 bit.

Reply:
According to the langref the backend can widen the store up to the alignment, if not using sanitizers. It just doesn't today

Reply:
Even if we can prove it doesn't trap, it's still not legal to store to the trailing bytes; they could contain other data.  So we have to split the store.

Reply:
I guess this has to be a target option then

Reply:
So what is the conclusion here? Do we want to have a custom lowering for X86?

Reply:
On targets with SIMD vectors (basically any modern CPU target), a 12-byte store is going to be slower than a 16-byte store, so we don't want to generate a 12-byte store.  So clang has an ABI rule that says we can generate a 16-byte store instead.

If you have some target that actually wants 12-byte stores for some reason, it needs to be a target-specific ABI rule in clang: on that target, we just directly generate 12-byte stores, instead of 16-byte stores.

Reply:
Hmm, that sounds essentially what the option that was removed in this PR is doing.

Reply:
Does it make sense to move the logic to back end? If a target doesn't support odd-byte load/store, we legalize it.

Reply:
> that sounds essentially what the option that was removed in this PR is doing

Yes, exactly.

> Does it make sense to move the logic to back end

The way LLVM IR is defined, a `<3 x float>` stores 12 bytes, and we can't really change that.

Reply:
The ideal control would be more specific than this. We have legal 12-byte load/store, but not 3/6 or other odd sizes

Reply:
Then I guess this can be more general: add a virtual target specific function to return the next legal size given N. If the two numbers don't match, we can extend the handling here to support "arbitrary" N.

Reply:
I'd expect this to be in terms of type, not size but yes.

Reply:
I took another look and this is not only a target option. For example, if it is hlsl, no matter what target it is (spir-v or dxil), vec3 is preserved. I tried to add a function `TargetInfo::getOptimalVectorNumElements(const llvm::FixedVectorType *)`, but it can't cover the language related stuff. If we add an extra argument, like `bool IsHLSL)`, it is too ugly.

Reply:
Eventually we might do it like:

```
bool preserveVec3(Triple TT, bool IsHLSL) {
  if (IsHLSL)
    return true;
  if (TT.isAMDGPU())
    return true;
  return false;
}
```

Reply:
I don't follow. The language merely states if the type has the padding bytes. It doesn't mandate you optimize down to the used elements

Reply:
Don't see what's so ugly. The target hook (which should be a function of the type) would be aware of the language options

Reply:
`unsigned getOptimalVectorNumElements(const llvm::FixedVectorType *, const LangOptions &)`?

Reply:
I would return the VectorType to use instead of doing it in terms of elements. Maybe the input should be the QualType?